tools/topology/topology2/platform/intel/speaker-echo-ref.conf

```diff
@@ -14,12 +14,26 @@ Object.Pipeline {
 				stream_name	$SPEAKER_CODEC_NAME
 				node_type	$I2S_LINK_INPUT_CLASS
 				num_input_audio_formats 1
-				Object.Base.input_audio_format [
-					{
-						in_bit_depth		32
-						in_valid_bit_depth	32
+				IncludeByKey.SSP_SPK_FMT_24 {
+				"true" {
+					Object.Base.input_audio_format [
+						{
+							in_bit_depth		32
+							in_valid_bit_depth	24
+							in_sample_type		$SAMPLE_TYPE_MSB_INTEGER
+							in_fmt_cfg		"$[($in_channels | ($in_valid_bit_depth * 256))]"
+						}
+					]
 					}
-				]
+				"false" {
+					Object.Base.input_audio_format [
+						{
+							in_bit_depth		32
+							in_valid_bit_depth	32
+						}
+					]
+					}
+				}
```

Comment:
... but here you do have a 'false' case. So that makes me wonder if this whole patch is consistent.

I guess we'll have to ask @ranj063 for feedback, this whole format selection remains a mystery to the rest of us....

Reply:
The same overwriting input, output formats with 24 bits/32 bits from io-gateway-capture.conf. The patch is consistent.

Reply:
@plbossart I think that depends on whether the original code set the audio format or not. If audio_format is set in the original code, then we should have a false case to set the audio_format as before. If not, we don't need to set the the audio_format since default audio_format worked before. Make sense?

Reply:
Can we please go back to the problem statemetn

"use 24 bits format on playback and capture pipelines if codec needs, e.g. RT5650 HS/SPK codec."

What does the codec need, if as I read "codec dai no need to change the sample bits 24 bits"

Reply:
@plbossart , 
> What does the codec need, if as I read "codec dai no need to change the sample bits 24 bits"

Sorry for the confusion. My description is not so precisely. I mean no need to change additional attribute  "sample_bits" in dai.ssp.  It will influence the bclk freq. 

Detail problem statement:
#aplay -r 4800 -c 2 -f s32 /dev/zero
snd_pcm_hw_params: Invalid argument  because the rt5645/50 codec not support S32_LE, but support S24_LE.

kernel: [ 9845.213552] rt5645 i2c-10EC5650:00: ASoC: error at snd_soc_dai_hw_params on rt5645-aif2: -22

That is because default DSP/Copier format is 32 valid bits in 32 bits container transmission to SSP. In order to support other specific format codec, we config DSP/Copier as 24 valid bits(LSB) in 32 bits container. By sending the valid_bits ipc token, ipc4 can process and pass S24_LE request to backend codec. In the end, rt5650 receives the 24 bits data from SSP bus.

static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
                                       struct snd_pcm_hw_params *params)
{
...
                valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(ipc4_fmt->fmt_cfg);
                dev_dbg(component->dev, "Set %s to %d bit format\n", dai->name, valid_bits);
        }

        /* Set format if it is specified */
        switch (valid_bits) {
        case 16:
                snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S16_LE);
                break;
        case 24:
                snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S24_LE);
                break;
        case 32:
                snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
                break;
        default:
                break;
        }

        switch (ipc4_copier->dai_type) {
        case SOF_DAI_INTEL_SSP:
                ipc4_ssp_dai_config_pcm_params_match(sdev, (char *)rtd->dai_link->name, params);
                break;